File: src/leet/easy/findifpathexistsingraph.rs
use std::collections::{HashSet, VecDeque};
pub struct Solution;

impl Solution {
    pub fn valid_path(n: i32, edges: Vec<Vec<i32>>, source: i32, destination: i32) -> bool {
        let mut adjacency_list = vec![vec![]; n as usize];

        for edge in &edges {
            let a = edge[0] ;
            let b = edge[1];
            
            adjacency_list[a as usize].push(b);
            adjacency_list[b as usize].push(a);
        }
        let mut seen: HashSet<i32> = HashSet::new();
        let mut queue: VecDeque<i32> = VecDeque::with_capacity(n as usize);
        
        queue.push_back(source);
        seen.insert(source);
 
        while let Some(node) = queue.pop_front() {
            if node == destination {
                return true;
            }

            for &neighbor in &adjacency_list[node as usize] {
                if !seen.contains(&neighbor) {
                    seen.insert(neighbor);
                    queue.push_back(neighbor);
                }
            }
        }
        
        return false;
    }
}

File: leet/easy/findifpathexistsingraph.py
# https://leetcode.com/problems/find-if-path-exists-in-graph
from collections import deque
from typing import List


class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        adjacency_list = [[] for i in range(n)]

        for edge in edges:
            adjacency_list[edge[0]].append(edge[1])
            adjacency_list[edge[1]].append(edge[0])

        queue = deque([source])
        visited = {source}

        while queue:
            node = queue.popleft()

            if node == destination:
                return True

            for neighbour in adjacency_list[node]:
                if neighbour not in visited:
                    visited.add(neighbour)

                    queue.append(neighbour)
        return False
